[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1QX1--2YPPGQAMWXxhGPS8V9MHr1nTBem?usp=sharing)

Submission by: Chiara Vega

# Recommendation Systems

A recommendation system helps users find compelling content in a large corpora. For example, the Google Play Store provides millions of apps, while YouTube provides billions of videos. More apps and videos are added every day. How can users find new and compelling content? Yes, one can use search to access content. However, a recommendation engine can display items that users might not have thought to search for on their own.

In this assignment you will use the [MovieLens dataset](https://grouplens.org/datasets/movielens/latest) called `ml-latest`. You are free to experiment and develop your code with the `ml-latest-small` dataset but you need to **submit** your work with the `ml-latest` dataset which is much larger and will lead to better results. The `ml-latest` dataset consists of several files:

* `genome-scores.csv`: a relevance score that a particular movie has a particular tag. Tags are things like "action", "romance", "violence", etc. The relevance score ranges from 0 to 1. The higher the score, the more relevant the tag is to the movie.
* `genome-tags.csv`: the tag name for each tag id.
* `links.csv`: the id of the movie on MovieLens, the id of the movie on IMDB, and the id of the movie on TMDB.
* `movies.csv`: the id of the movie on MovieLens, the title of the movie, and a list of genres associated with the movie.
* `ratings.csv`: the id of the movie on MovieLens, the id of the user rating the movie, the rating, and the timestamp of the rating.
* `tags.csv`: the id of the movie on MovieLens, the id of the user tagging the movie, the tag, and the timestamp of the tag.



## Task 1: SVD for Recommendations (30 points)

Go over the [recommendation system overview](https://datajobs.com/data-science-repo/Recommender-Systems-%5BNetflix%5D.pdf) and write below what matrix factorization can offer and what challenges it faces. To do so, consult the [Surprise](http://surpriselib.com/) library and choose the SVD API of `surpise.prediction_algorithms`. Describe all equations involved and why they turn out to be the way they are.

NOTE1: In this task you are expected to do some reading and consultation of APIs. Be as descriptive as possible so that a novice computer scientist can understand your answer. There is no minimum of max "page limit". Pay particular attention to [this reference](https://sifter.org/~simon/journal/20061211.html)

NOTE2: Although we have recommended the Surprise library, it is worth mentioning that Microsoft has a [recommender system toolkit](https://github.com/microsoft/recommenders). Since recommendation systems are almost everywhere you look in commercial / consumer applications, you may want to spend the time after this assignment to learn more about the field.

**Write your answer here or in a separate markdown file + images**

### Matrix Factorization Overview

Matrix Factorization is a powerful technique that offers remarkable capabilities in Recommendation Systems.


### What Matrix Factorization Offers
Through analysis of patterns and hidden features, matrix factorization enables to predictions of affinities between the users and items and then providing tailored suggestions. In other words, matrix factorization can unravel user preferences and item characteristics, enabling personalized recommendations. Another strong point of matrix factorization is good scalability to handle massive datasets.

---

### Challenges that Matrix Factorization Faces
However, matrix factorization encounters challenges. When dealing with new users and/or items that lack sufficient historical data, matrix factorization faces a "cold start" problem due to its inability to address the new information. Effective approaches must be devised to provide recommendations without extensive user/item information.

---

### Prediction
Let's move on to talk about the equations that drive matrix factorization's predictive capabilities by going over its prediction equation. Let each item $i$ be associated with vector $q_i \in \mathbb{R}$ and each user $u$ be associated with a vector $p_u \in \mathbb{R}$. The dot product $q_i^Tp_u$ captures the relationship between the user and the object (i.e. the user's general interest in the item). Let $r_{ui}$ denote user's rating of item, which is then estimated by:

$\hat{r}_{ui} = q_i^Tp_u$

Now in the case of the popular Singular Value Decomposition (SVD) algorithm, as popularized by Simon Funk during the Netflix Prize competition, the prediction equation is as follows:

$\hat{r}_{ui} = \mu + b_i + b_u + q_i^Tp_u$ where $\mu$ represents the overall average rating, $b_i$ the bias term of $i$, and $b_u$ the bias term of $u$.

The SVD prediction equation estimates the rating based on user biases, item biases, and latent factor vectors - allowing for prediction of ratings for user-item pairs not explicitly rated previously.

---

### Optimization
We move onto the optimization of the algorithm that minimizes the difference between the predicted $\hat{r}_{ui}$ and ground truth $r_{ui}$. The minimization incorporates a regularized squared error for estimating all unknown, and the unbiased equation is as follows:

$\min_\limits{q^*, p^*} \sum\limits_{(u,i) \in \kappa} (r_{ui} - q_i^Tp_u)^2 + \lambda (||q_i||^2 + ||p_u||^2)$ where $\kappa$ is the training set - the set of ($u,i$) pairs for which $r_{ui}$ is known.

The biased optimization equation is as follows:

$\min_\limits{q^*, p^*,b^*} \sum\limits_{(u,i) \in \kappa} (r_{ui} - \mu - b_u - b_i - q_i^Tp_u)^2 + \lambda (||p_u||^2 + ||q_i||^2 + b_u^2 + b_i^2)$

The model adjusts the latent factor vectors and biases to find the optimal configuration that provides a good balance between prediction accuracy and model complexity.

---

### Surprise Library - SVD API

The Surprise library provides a convenient SVD API (surprise.prediction_algorithms.matrix_factorization.SVD) implementing the above equations for exploring matrix factorization - taking care of the underlying computations. Though implementation details may vary slightly, the principles remain the same, not to mention, one is provided methods to train the model, make predictions, and evaluate model performance.

---

### Conclusion

To summarize, matrix factorization offers the mathematical framework to understand user-item interactions and make recommendations. The equations involved capture the various aspects including ratings, biases, and regularization. And finally, Surprise library's SVD API simplifies implementation of matrix factorization for building recommender systems.

## Task 2: Implement the SVD based matrix facrorization algorithm (40 points)

Implement the matrix factorization algorithm on the movielens data using the surprise API. You need to provide a clear explanation of the code you wrote and you need to report and explain the recommendation metric on a test dataset.

In [ ]:
# Task 2 code starts here

Note: I will be using the [MovieLens dataset](https://grouplens.org/datasets/movielens/latest) called `ml-latest` as indicated in Task 1 - containing 27 million ratings.

In [ ]:
!pip install surprise

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 772.0/772.0 kB 11.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.3-cp310-cp310-linux_x86_64.whl size=3095456 sha256=ed3adad63962db195bd471676a8a0dd0b4aed715bb59d5ad11e23d95e7545c61
  Stored in directory: /root/.cache/pip/wheels/a5/ca/a8/4e28def53797fdc4363ca4af740db15a9c2f1595ebc51fb445
Successfully built scikit-surprise


In [ ]:
# import necessary libraries
import numpy as np
import pandas as pd

from surprise import Reader, SVD, Dataset, accuracy
from surprise.model_selection import train_test_split

In [ ]:
# load and preview movies.csv file
df_movies = pd.read_csv('movies.csv')
df_movies.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [ ]:
# check for missing values
df_movies.isna().sum()

movieId    0
title      0
genres     0
dtype: int64

In [ ]:
# load and preview ratings.csv file
df_ratings = pd.read_csv('ratings.csv')
df_ratings.head()

userId  movieId  rating   timestamp
0       1      307     3.5  1256677221
1       1      481     3.5  1256677456
2       1     1091     1.5  1256677471
3       1     1257     4.5  1256677460
4       1     1449     4.5  1256677264

In [ ]:
# check for missing values
df_ratings.isna().sum()

userId       0
movieId      0
rating       0
timestamp    0
dtype: int64

In [ ]:
#merge movies.csv and ratings.csv into a new dataframe
merged_df = pd.merge(df_movies, df_ratings, on="movieId")
merged_df.head()

movieId             title                                       genres  \
0        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
1        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
2        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
3        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
4        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   

   userId  rating   timestamp  
0       4     4.0  1113765937  
1      10     5.0   948885850  
2      14     4.5  1442169375  
3      15     4.0  1370810063  
4      22     4.0  1237622631

In [ ]:
merged_df.isna().sum()

movieId      0
title        0
genres       0
userId       0
rating       0
timestamp    0
dtype: int64

In [ ]:
# combine and preview movieId and title
merged_df['movie'] = merged_df['movieId'].map(str) + str(': ') + merged_df['title'].map(str)
merged_df.head()

movieId             title                                       genres  \
0        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
1        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
2        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
3        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
4        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   

   userId  rating   timestamp                movie  
0       4     4.0  1113765937  1: Toy Story (1995)  
1      10     5.0   948885850  1: Toy Story (1995)  
2      14     4.5  1442169375  1: Toy Story (1995)  
3      15     4.0  1370810063  1: Toy Story (1995)  
4      22     4.0  1237622631  1: Toy Story (1995)

In [ ]:
# get the lowest and highest rating
min_rating, max_rating = merged_df['rating'].min(), merged_df['rating'].max()

# set up Surprise's Reader object
reader = Reader(rating_scale=(min_rating, max_rating))

# load the data into Surprise's Dataset object
# movie, userId, and rating as the essentials for collaborative filtering
data = Dataset.load_from_df(merged_df[['movie', 'userId', 'rating']], reader)

In [ ]:
# split data into training and testing sets
# 70% training and 30% test
trainset, testset = train_test_split(data, test_size=0.3, random_state=123)

In [ ]:
# build SVD baseline model
model = SVD()

# train the baseline model on the training set
model.fit(trainset)

# make predictions on the test set
predictions = model.test(testset)

# RMSE as recommendation metric
rmse = accuracy.rmse(predictions)
print("RMSE (test set):", rmse)

RMSE: 0.7967
RMSE (test set): 0.7966706960814334


RMSE (Root Mean Square Error) is a commonly used metric for recommendation systems. RMSE is typically used for regression problems where it measures the average magnitude of the differences between the predicted scalar value (in this case the predicted ratings) and true/ground truth scalar value (in this case actual ratings) for a given data point (in this case the users).

## Task 3: Perform hyperparameter tuning (30 points)

In this task, you will borrow candidate hyperparameter values from [here](https://github.com/microsoft/recommenders/blob/main/examples/04_model_select_and_optimize/nni_surprise_svd.ipynb) but implement a simple random search to tune them. You can use [Optuna](https://optuna.org/) for such exercise. You need to report the best hyperparameters and the corresponding recommendation metric on a test dataset.

In [ ]:
# Task 3 code starts here

In [ ]:
!pip install optuna

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 390.6/390.6 kB 7.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 224.5/224.5 kB 25.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.7/78.7 kB 10.6 MB/s eta 0:00:00


In [ ]:
# import necessary packages
import optuna
from optuna.samplers import RandomSampler

In [ ]:
# define objective function for hyperparameter space
def objective(trial):
    params = {
        'n_factors': trial.suggest_int('n_factors', 10, 50),
        'n_epochs': trial.suggest_int('n_epochs', 10, 50),
        'lr_all': trial.suggest_uniform('lr_all', 0.01, 0.1),
        'reg_all': trial.suggest_uniform('reg_all', 0.01, 0.1)
    }

    # build model with the optimal hyperparameters
    model_svd = SVD(**params)

    # train the baseline model on the training set
    model_svd.fit(trainset)

    # make predictions on the test set
    predictions = model_svd.test(testset)

    # RMSE as recommendation metric
    rmse = accuracy.rmse(predictions)

    return rmse

In [ ]:
# perform a simple random search using Optuna
study = optuna.create_study(direction='minimize', sampler=RandomSampler())
study.optimize(objective, n_trials=10)

# print the best hyperparameters and RMSE recommendation metric
best_params = study.best_params
best_rmse = study.best_value
print("Best Hyperparameters:", best_params)
print("Best RMSE:", best_rmse)

[I 2023-06-21 20:03:57,360] A new study created in memory with name: no-name-5dd9cca5-8a07-48df-9c40-f3c0a9befec5
<ipython-input-26-a760a46d3d16>:6: FutureWarning: suggest_uniform has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use suggest_float instead.
  'lr_all': trial.suggest_uniform('lr_all', 0.01, 0.1),
<ipython-input-26-a760a46d3d16>:7: FutureWarning: suggest_uniform has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use suggest_float instead.
  'reg_all': trial.suggest_uniform('reg_all', 0.01, 0.1)
[I 2023-06-21 20:08:45,820] Trial 0 finished with value: 0.832502978600477 and parameters: {'n_factors': 20, 'n_epochs': 15, 'lr_all': 0.039865007599820125, 'reg_all': 0.06475569339673477}. Best is trial 0 with value: 0.832502978600477.


RMSE: 0.8325


[I 2023-06-21 20:20:45,342] Trial 1 finished with value: 0.8853341884505642 and parameters: {'n_factors': 41, 'n_epochs': 39, 'lr_all': 0.09320634551338079, 'reg_all': 0.0544815306714228}. Best is trial 0 with value: 0.832502978600477.


RMSE: 0.8853


[I 2023-06-21 20:24:59,671] Trial 2 finished with value: 0.8497839473583463 and parameters: {'n_factors': 10, 'n_epochs': 13, 'lr_all': 0.05727501937309631, 'reg_all': 0.05145298389967878}. Best is trial 0 with value: 0.832502978600477.


RMSE: 0.8498
RMSE: 0.8465


[I 2023-06-21 20:33:01,756] Trial 3 finished with value: 0.8465045914065309 and parameters: {'n_factors': 12, 'n_epochs': 33, 'lr_all': 0.041288098335264534, 'reg_all': 0.012262031172745116}. Best is trial 0 with value: 0.832502978600477.
[I 2023-06-21 20:43:48,522] Trial 4 finished with value: 0.8586120035404373 and parameters: {'n_factors': 36, 'n_epochs': 37, 'lr_all': 0.06527352316712129, 'reg_all': 0.08208019562453965}. Best is trial 0 with value: 0.832502978600477.


RMSE: 0.8586


[I 2023-06-21 20:57:37,880] Trial 5 finished with value: 0.8405899525392451 and parameters: {'n_factors': 42, 'n_epochs': 48, 'lr_all': 0.04946948694068024, 'reg_all': 0.0746185099063303}. Best is trial 0 with value: 0.832502978600477.


RMSE: 0.8406


[I 2023-06-21 21:02:29,764] Trial 6 finished with value: 0.8416407391451362 and parameters: {'n_factors': 47, 'n_epochs': 12, 'lr_all': 0.0549145953366542, 'reg_all': 0.05403163548074948}. Best is trial 0 with value: 0.832502978600477.


RMSE: 0.8416
RMSE: 0.9601


[I 2023-06-21 21:07:10,240] Trial 7 finished with value: 0.9600742117838735 and parameters: {'n_factors': 45, 'n_epochs': 11, 'lr_all': 0.09774344041730695, 'reg_all': 0.01774202417576652}. Best is trial 0 with value: 0.832502978600477.


RMSE: 0.8095


[I 2023-06-21 21:12:27,750] Trial 8 finished with value: 0.8095254174270777 and parameters: {'n_factors': 27, 'n_epochs': 16, 'lr_all': 0.02378951396535218, 'reg_all': 0.0480741383143852}. Best is trial 8 with value: 0.8095254174270777.


RMSE: 0.8158


[I 2023-06-21 21:18:15,479] Trial 9 finished with value: 0.8158001555220075 and parameters: {'n_factors': 24, 'n_epochs': 19, 'lr_all': 0.01756027969303375, 'reg_all': 0.06329173456538723}. Best is trial 8 with value: 0.8095254174270777.


Best Hyperparameters: {'n_factors': 27, 'n_epochs': 16, 'lr_all': 0.02378951396535218, 'reg_all': 0.0480741383143852}
Best RMSE: 0.8095254174270777


### Best Hyperparameters

'n_factors': 27

'n_epochs': 16

'lr_all': 0.02378951396535218

'reg_all': 0.0480741383143852

### Best RMSE

(corresponding recommendation metric chosen)

RMSE: 0.8095254174270777